editors/code/src/installation/download_file.ts

```diff
@@ -14,31 +21,31 @@ export async function downloadFile(
     destFilePermissions: number,
     onProgress: (readBytes: number, totalBytes: number) => void
 ): Promise<void> {
-    const res = await fetch(url);
+    const res = await fetch(url).catch(DownloadFileError.rethrow("Failed at initial fetch"));
 
     if (!res.ok) {
         console.log("Error", res.status, "while downloading file from", url);
         console.dir({ body: await res.text(), headers: res.headers }, { depth: 3 });
 
-        throw new Error(`Got response ${res.status} when trying to download a file`);
+        throw new DownloadFileError(`Got response ${res.status}`);
     }
 
     const totalBytes = Number(res.headers.get('content-length'));
     assert(!Number.isNaN(totalBytes), "Sanity check of content-length protocol");
 
+    console.log("Downloading file of", totalBytes, "bytes size from", url, "to", destFilePath);
+
     let readBytes = 0;
+    res.body.on("data", (chunk: Buffer) => {
+        readBytes += chunk.length;
+        onProgress(readBytes, totalBytes);
+    });
 
-    console.log("Downloading file of", totalBytes, "bytes size from", url, "to", destFilePath);
+    const destFileStream = fs.createWriteStream(destFilePath, { mode: destFilePermissions });
 
-    return new Promise<void>((resolve, reject) => res.body
-        .on("data", (chunk: Buffer) => {
-            readBytes += chunk.length;
-            onProgress(readBytes, totalBytes);
-        })
-        .on("error", reject)
-        .pipe(fs
-            .createWriteStream(destFilePath, { mode: destFilePermissions })
-            .on("close", resolve)
-        )
-    );
+    await pipeline(res.body, destFileStream).catch(DownloadFileError.rethrow("Piping file error"));
+    return new Promise<void>(resolve => {
```

Comment:
Why do we need a Promise here at all? I would imagine that the following would work (with some exception safety added just in case):

```
try {
    await pipeline(res.body, destFileStream);
    return;
} finally {
    destFileStream.destory()
}
```

Reply:
Unfortunately it doesn't work on my laptop on windows ;( We still need to await the close event... Please see that discussion you originally referred to.

Reply:
Oh, sorry, I've missed the last bit of the discussion :( Yeah, I guess the current code is what we should do then.

Reply:
By the way, regarding `destroy()` in finally, this is a bit convoluted...
The streams do get automatically disposed on `"error"` event. [See documentation](https://nodejs.org/api/stream.html#stream_stream_pipeline_streams_callback)
> stream.pipeline() will call stream.destroy(err) on all streams except:
>
> Readable streams which have emitted 'end' or 'close'.
> Writable streams which have emitted 'finish' or 'close'.